pyro/poutine/trace_struct.py

```diff
@@ -145,17 +148,22 @@ def compute_log_prob(self, site_filter=lambda name, site: True):
         """
         for name, site in self.nodes.items():
             if site["type"] == "sample" and site_filter(name, site):
-                try:
-                    site["log_prob"]
-                except KeyError:
-                    log_p = site["fn"].log_prob(site["value"], *site["args"], **site["kwargs"])
-                    site["unscaled_log_prob"] = log_p
-                    log_p = scale_and_mask(log_p, site["scale"], site["mask"])
-                    site["log_prob"] = log_p
-                    site["log_prob_sum"] = log_p.sum()
-                    if is_validation_enabled():
-                        warn_if_nan(site["log_prob_sum"], "log_prob_sum at site '{}'".format(name))
-                        warn_if_inf(site["log_prob_sum"], "log_prob_sum at site '{}'".format(name), allow_neginf=True)
+                if "log_prob" not in site:
+                    try:
+                        log_p = site["fn"].log_prob(site["value"], *site["args"], **site["kwargs"])
```

Comment:
nit: I think we should only catch the exception in `.log_prob` by wrapping it inside `try..except`. Also, I was mistaken that `log_prob_sum` was calling `compute_log_prob`, so we should also do it inside `log_prob_sum`. 😄Looks great otherwise.

```python
try:
    log_p = site["fn"].log_prob
except ValueError:
    _, exc_value, traceback = sys.exc_info()
    ...
...
```

Reply:
Could you also please add this nice logic to `compute_score_parts` to that we can catch errors in guides?